Performing LES on devices

Shiyu-Sandy-Du opened the pull request:
In this PR,

- ax_helm_full_device is implemented for hip and cuda
- sigma and vreman SGS models are implemented for hip and cuda as well
- bug is fixed in the pnpn pressure res computation for full stress formulation on devices

Some other things to be discussed:

- [x] In pnpn_res_stress_device.F90, device math is called for multiple times. If we merge those kernels together as much as possible, how much can we benefit from it considering the solving time for the pressure equation is not short?
- [x] Implementation for 1D in ax_helm_full_device has not been implemented and thus autotune is not yet implemented as well. But do we really need it? I cannot see it in the vector version of ax_helm_device implementation. Is the 1D or kstep discussion for vector version of ax_helm happened before?

timofeymukha commented:
I'll leave it to the GPU experts to review, just gonna say -- very impressive work, Shiyu!

njansson commented:
This must be synced with develop before we can consider a merge ( due to all the dependency changes)

Shiyu-Sandy-Du commented:
> This must be synced with develop before we can consider a merge ( due to all the dependency changes)

Exactly, I am working on the conflicts. And wait PR #1581 to be merged for a proper git push

njansson commented:
Nice!

njansson commented on src/les/bcknd/device/hip/hip_sigma_nut.f90:
Should we declare these interface in `device_sigma_nut.F90`, as with our other device implementations? One motivation would be to avoid the multiple inclusion warnings with CCE

Shiyu-Sandy-Du replied:
Yes it should be done.

njansson commented on src/les/bcknd/device/hip/hip_vreman_nut.f90:
Same comment as for `hip_sigma_nut`

njansson commented:
To answer some of the questions,

> * [ ]  In pnpn_res_stress_device.F90, device math is called for multiple times. If we merge those kernels together as much as possible, how much can we benefit from it considering the solving time for the pressure equation is not short?

There are gains to be made here, by fusing the e.g. the cluster of col3 and sub2. Remember that each call to a device math functions comes with a launch latency, which is not negligible.

> * [ ]  Implementation for 1D in ax_helm_full_device has not been implemented and thus autotune is not yet implemented as well. But do we really need it? I cannot see it in the vector version of ax_helm_device implementation. Is the 1D or kstep discussion for vector version of ax_helm happened before?

Exactly, there's no 1d version of these, since 1d will run out of shared memory for most polynomial orders.

MartinKarp commented on src/les/bcknd/device/vreman_device.f90:
```diff
@@ -0,0 +1,136 @@
+    call dudxyz (a32%x, w%x, coef%drdy, coef%dsdy, coef%dtdy, coef)
+    call dudxyz (a33%x, w%x, coef%drdz, coef%dsdz, coef%dtdz, coef)
+
+    call coef%gs_h%op(a11%x, a11%dof%size(), GS_OP_ADD)
```
I think you can just replace this with `gs_h%op(a11, GS_OP_ADD)` for example

MartinKarp commented on src/les/bcknd/device/cuda/sigma_nut_kernel.h:
```diff
@@ -0,0 +1,183 @@
+    T pi_3;
+    T tmp1;
+    
+    pi_3 = 4.0/3.0*atan(1.0);
```
dont we have pi somewhere here?

Shiyu-Sandy-Du replied:
Do we have it defined somewhere else?

njansson replied:
it can be fetched from the header files

njansson replied:
Given that this is a templated variable, I would say that we leave it for later to fix.

MartinKarp reviewed (approved):
Great stuff Shiyu! I just added two comments, feel free to take them to heart or not. 

One thing I noted is the use of quite a lot of math functions such as pow, divisions and so on. I think we should just be aware this might be an issue if people run in single precision and one can perhaps use xreal in some places then. This is just a note for the future though.

Shiyu-Sandy-Du commented:
> Great stuff Shiyu! I just added two comments, feel free to take them to heart or not.
> 
> One thing I noted is the use of quite a lot of math functions such as pow, divisions and so on. I think we should just be aware this might be an issue if people run in single precision and one can perhaps use xreal in some places then. This is just a note for the future though.

I changed the unnecessary pow into multiplications here. And I think we should note the sin, cos and sqrt here for future maybe.
